[Feature] [kubectl-plugin] Expose setting `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` in ray job submit

CheyuWu opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Currently shutdownAfterJobFinishes and ttlSecondsAfterFinished aren't being exposed in kubectl ray job submit, although they are available in the [CRD](https://docs.ray.io/en/latest/cluster/kubernetes/getting-started/rayjob-quick-start.html#rayjob-configuration).
<!-- Please give a short summary of the change and the problem this solves. -->

### Ray Operator

- Add support for `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` in RayJob spec.

- Implement validation logic: if `shutdownAfterJobFinishes` is False, then `ttlSecondsAfterFinished` must be 0 or unset.

- Add corresponding unit tests.

### kubectl-plugin
- Add `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` flags to the rayjob submit command.

- Implement validation logic: if `shutdownAfterJobFinishes` is False, then `ttlSecondsAfterFinished` must be 0 or unset.

- Add corresponding unit tests.

## Related issue number
Closes #3560 
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [ ] Manual tests
  - [ ] This PR is not tested :(

CheyuWu commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -40,40 +40,42 @@ const (
+	noWait                   bool
+	dryRun                   bool
+	verbose                  bool
+	shutdownAfterJobFinishes bool
+	ttlSecondsAfterFinished  int32
```
Add `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished`

CheyuWu commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
 			options.runtimeEnvJson = string(runtimeJson)
 		}
+
+		if !options.RayJob.Spec.ShutdownAfterJobFinishes && options.RayJob.Spec.TTLSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
+		}
+
```
Implement validation logic when `--filename` is provided
> if shutdownAfterJobFinishes is False, then ttlSecondsAfterFinished must be 0 or unset.

CheyuWu commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
+
 	} else if strings.TrimSpace(options.rayjobName) == "" {
 		return fmt.Errorf("Must set either yaml file (--filename) or set Ray job name (--name)")
+	} else if options.fileName == "" {
+		if !options.shutdownAfterJobFinishes && options.ttlSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
+		}
```
Implement validation logic when `--filename` is not provided
> if shutdownAfterJobFinishes is False, then ttlSecondsAfterFinished must be 0 or unset.

CheyuWu commented on kubectl-plugin/pkg/cmd/job/job_submit_test.go:
```diff
@@ -30,7 +30,7 @@ func TestRayJobSubmitComplete(t *testing.T) {
 }
 
-func TestRayJobSubmitValidate(t *testing.T) {
+func TestRayJobSubmitWithYamlValidate(t *testing.T) {
```
Split TestRayJobSubmitValidate into two test cases: one for YAML-based submission and one for non-YAML-based submission.

CheyuWu commented on kubectl-plugin/pkg/cmd/job/job_submit_test.go:
```diff
@@ -99,6 +122,52 @@ spec:
 	}
 }
 
+func TestRayJobSubmitWithoutYamlValidate(t *testing.T) {
```
Split TestRayJobSubmitValidate into two test cases: one for YAML-based submission and one for non-YAML-based submission.

CheyuWu commented on ray-operator/controllers/ray/utils/validation.go:
```diff
@@ -144,6 +144,10 @@ func ValidateRayJobSpec(rayJob *rayv1.RayJob) error {
 		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false is not allowed to be suspended")
 	}
 
+	if !rayJob.Spec.ShutdownAfterJobFinishes && rayJob.Spec.TTLSecondsAfterFinished != 0 {
+		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false cannot have TTLSecondsAfterFinished is non-zero")
+	}
+
```
Implement validation logic in ray operator
> if shutdownAfterJobFinishes is False, then ttlSecondsAfterFinished must be 0 or unset.

CheyuWu commented:
Hi @MortalHappiness and @kevin85421  PTAL

davidxia commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
 			options.runtimeEnvJson = string(runtimeJson)
 		}
+
+		if !options.RayJob.Spec.ShutdownAfterJobFinishes && options.RayJob.Spec.TTLSecondsAfterFinished != 0 {
```
unrelated to this PR, but when I was testing this change, I ran `kubectl ray job submit -f ray-job.sample.yaml --working-dir . --shutdown-after-job-finishes --ttl-seconds-after-finished 600 -- python /home/ray/samples/sample_code.py` and wondered why the new flags weren't used. I had to look at this source to understand `-f/--filename` makes the command ignore many of these other flags.

MortalHappiness replied:
+1. I think we should take the filename as a base, and flags will override the values in the file.

I created an issue to track this https://github.com/ray-project/kuberay/issues/3639

davidxia replied:
The behavior of `kubectl ray create cluster` is to error with an actionable error message if declarative file and CLI flags are used together. We can consider doing that here too if we think it's not worth it to have override logic.

CheyuWu replied:
It is worth overriding the values in the file. I will change this.

davidxia commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
 		return fmt.Errorf("Must set either yaml file (--filename) or set Ray job name (--name)")
+	} else if options.fileName == "" {
+		if !options.shutdownAfterJobFinishes && options.ttlSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
```
```suggestion
			return fmt.Errorf("--ttl-seconds-after-finished is only supported when --shutdown-after-job-finishes is set to true")
```

davidxia commented on ray-operator/controllers/ray/utils/validation.go:
```diff
@@ -144,6 +144,10 @@ func ValidateRayJobSpec(rayJob *rayv1.RayJob) error {
 	}
 
+	if !rayJob.Spec.ShutdownAfterJobFinishes && rayJob.Spec.TTLSecondsAfterFinished != 0 {
+		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false cannot have TTLSecondsAfterFinished is non-zero")
```
```suggestion
		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false cannot have TTLSecondsAfterFinished")
```

davidxia commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
 	} else if strings.TrimSpace(options.rayjobName) == "" {
 		return fmt.Errorf("Must set either yaml file (--filename) or set Ray job name (--name)")
+	} else if options.fileName == "" {
+		if !options.shutdownAfterJobFinishes && options.ttlSecondsAfterFinished != 0 {
```
also check ttlSecondsAfterFinished ≥ 0 in both cases? Right now this command will create a RayJob with ttlSecondsAfterFinished: -10

```
kubectl ray --context gke_kubeflow-platform_europe-west4-b_ml-compute-1 -n hyperkube job submit --name dxia-test --shutdown-after-job-finishes --ttl-seconds-after-finished -10 --working-dir . --dry-run  -- python /home/ray/samples/sample_code.py | grep -B 1 ttlSeconds
  submissionMode: InteractiveMode
  ttlSecondsAfterFinished: -10
```

CheyuWu replied:
I will fix this, thx

MortalHappiness commented on apiserver/pkg/model/converter_test.go:
```diff
@@ -303,9 +303,10 @@ var JobNewClusterTest = rayv1api.RayJob{
-		TTLSecondsAfterFinished: secondsValue,
-		RayClusterSpec:          &ClusterSpecTest.Spec,
+		RuntimeEnvYAML:           "mytest yaml",
+		ShutdownAfterJobFinishes: true,
```
Why change apiserver tests?

CheyuWu replied:
Because I have modified the ray operator that `ShutdownAfterJobFinishes` should be true; otherwise, `ttlSecondsAfterFinishedcannot`  cannot be set. 
[reference](https://docs.ray.io/en/latest/cluster/kubernetes/getting-started/rayjob-quick-start.html#rayjob-configuration)

MortalHappiness replied:
Seems unrelated to this PR? Could you do apiserver-related changes in a separate PR?

MortalHappiness commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -171,6 +173,8 @@ func NewJobSubmitCommand(cmdFactory cmdutil.Factory, streams genericclioptions.I
 	cmd.Flags().BoolVar(&options.dryRun, "dry-run", false, "print the generated YAML instead of creating the cluster. Only works when filename is not provided")
 	cmd.Flags().BoolVarP(&options.verbose, "verbose", "v", false, "Passing the '--verbose' flag to the 'ray job submit' command")
+	cmd.Flags().BoolVar(&options.shutdownAfterJobFinishes, "shutdown-after-job-finishes", false, "Shutdown the cluster after the job finishes")
+	cmd.Flags().Int32Var(&options.ttlSecondsAfterFinished, "ttl-seconds-after-finished", 0, "TTL seconds after finished. Only works when filename is not provided")
```
It still works when filename is provided right?

CheyuWu replied:
Currently, I will make sure it work in next commit.

CheyuWu commented:
Hi @davidxia, @MortalHappiness, PTAL
- Fixed the error message for better clarity.

- When using `--shutdown-after-job-finishes` or `--ttl-seconds-after-finished`, these flags will override the corresponding fields (`shutdownAfterJobFinishes` or `ttlSecondsAfterFinished`) in the file specified by `--filename`.

- Added validation to reject negative values for `ttlSecondsAfterFinished` or the `--ttl-seconds-after-finished` flag.

- Reverted changes related to the Ray Operator.

MortalHappiness commented on kubectl-plugin/pkg/cmd/job/job_submit_test.go:
```diff
@@ -71,6 +73,41 @@ spec:
+			expectError: "ttlSecondsAfterFinished is only supported when shutdownAfterJobFinishes is set to true",
+		},
+		{
+			name: "shutdownAfterJobFinishes is false and ttlSecondsAfterFinished is not zero",
```
```suggestion
			name: "shutdownAfterJobFinishes is true and ttlSecondsAfterFinished is not zero",
```

CheyuWu replied:
Ops,

kevin85421 reviewed (commented):
Why do we need to expose `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` at the same time? We can expose `ttlSecondsAfterFinished` only and configure `shutdownAfterJobFinishes` to true when `ttlSecondsAfterFinished` is set.

also cc @MortalHappiness @davidxia

davidxia commented:
ah that's true

MortalHappiness commented:
SGTM. @CheyuWu Could you only expose `ttlSecondsAfterFinished`? Thanks.

CheyuWu commented:
Hi @kevin85421 @MortalHappiness  @davidxia PTAL
I have updated it to expose only `ttlSecondsAfterFinished`.

## Manual Testing
> Set shutdownAfterJobFinishes to false and ttlSecondsAfterFinished to 2
> my_yaml.yaml
```yaml
apiVersion: ray.io/v1
kind: RayJob
metadata:
  name: rayjob-sample
spec:
  submissionMode: 'InteractiveMode'
  shutdownAfterJobFinishes: false 
  ttlSecondsAfterFinished: 2
  rayClusterSpec:
    rayVersion: '2.41.0'
    headGroupSpec:
      rayStartParams: {}
      template:
        spec:
          containers:
            - name: ray-head
              image: rayproject/ray:2.41.0
              ports:
                - containerPort: 6379
                  name: gcs-server
                - containerPort: 8265
                  name: dashboard
                - containerPort: 10001
                  name: client
              resources:
                limits:
                  cpu: "1"
                requests:
                  cpu: "200m"
    workerGroupSpecs:
      - replicas: 1
        minReplicas: 1
        maxReplicas: 5
        groupName: small-group
        rayStartParams: {}
        template:
          spec:
            containers:
              - name: ray-worker
                image: rayproject/ray:2.41.0
                resources:
                  limits:
                    cpu: "1"
                  requests:
                    cpu: "200m"

```

Only override ttlSecondsAfterFinished -> Success
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished 10 -- python entrypoint-python-sample.py
```

set `--ttl-seconds-after-finished` negative value -> failure
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished -10 -- python entrypoint-python-sample.py
```
```console
Error: --ttl-seconds-after-finished must be greater than or equal to 0
```

Not override yaml config
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml  -- python entrypoint-python-sample.py
```
```console
Error: ttlSecondsAfterFinished is only supported when shutdownAfterJobFinishes is set to true
```

CheyuWu commented on kubectl-plugin/pkg/cmd/job/job_submit.go:
```diff
@@ -40,42 +40,44 @@ const (
+	noWait                   bool
+	dryRun                   bool
+	verbose                  bool
+	shutdownAfterJobFinishes bool
```
Hi @MortalHappiness, If we remove `shutdownAfterJobFinishes`, it will cause an issue [here](https://github.com/ray-project/kuberay/pull/3627/files#diff-d7990c8ae1278889923e749905d9699559a9993a95a6bb7e7b62cca21968c3c7R144), where we won't be able to assign a value when generating the RayJob YAML.

If we try to determine `ShutdownAfterJobFinishes` based on the `ttlSecondsAfterFinished` field, there will be a problem when `ttlSecondsAfterFinished` is explicitly set to 0 (which is its default value).

Please let me know if I’ve misunderstood anything or missed any details.

MortalHappiness replied:
Ok there are some follow-up questions:

* Did you try setting `ttlSecondsAfterFinished` to `0` manually in the CLI, and does it work?
  * If yes, add a test that explicitly sets it to 0 and verify that it works.
  * If not, we can probably set `SubmitJobOptions.ttlSecondsAfterFinished` to `-1` in `Validate` if the user doesn't provide a value. Then in `Run`, we only set `ShutdownAfterJobFinishes` to true when the value is non-negative.

CheyuWu replied:
I ran the commands
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished 0 -- python entrypoint-python-sample.py
```
And it works
![image](https://github.com/user-attachments/assets/72241eff-7745-41e4-bd25-b07283e0fc99)

Also, 
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit --name rayjob-sample --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished 0 -- python entrypoint-python-sample.py
```
And it works
![image](https://github.com/user-attachments/assets/9a46018a-019b-4b12-adc7-b2825461321d)

CheyuWu replied:
I got an idea, we probably can set `ttl-seconds-after-finished` to -1 as default?

MortalHappiness replied:
It's ok if both setting it to `0` and omitting the flag work as expected. You don't need to set it to `-1`. You can preserve the `shutdownAfterJobFinishes` field. But could you add e2e tests for these two corner cases? Thanks.

CheyuWu replied:
OK, no problem

CheyuWu commented:
Hi @MortalHappiness PTAL

MortalHappiness commented on kubectl-plugin/test/e2e/kubectl_ray_job_submit_test.go:
```diff
@@ -100,6 +100,68 @@ var _ = Describe("Calling ray plugin `job submit` command on Ray Job", func() {
 		Expect(rayJob.Spec.RayClusterSpec.WorkerGroupSpecs[0].Template.Spec.NodeSelector["kubernetes.io/os"]).To(Equal("linux"))
 	})
+
+	It("successfully submits RayJob with TTL 0 using CLI flags only (no YAML config)", func() {
+		runtimeEnvFilePath := path.Join(kubectlRayJobWorkingDir, runtimeEnvSampleFileName)
+		cmd := exec.Command(
+			"kubectl", "ray", "job", "submit",
+			"--namespace", namespace,
+			"--name", rayJobName,
+			"--runtime-env", runtimeEnvFilePath,
+			"--ttl-seconds-after-finished", "0",
+			"--head-cpu", "1",
+			"--head-memory", "2Gi",
+			"--worker-cpu", "1",
+			"--worker-memory", "2Gi",
+			"--",
+			"python",
+			entrypointSampleFileName,
+		)
+		output, err := cmd.CombinedOutput()
+		Expect(err).NotTo(HaveOccurred())
+		// Retrieve the Job ID from the output
+		cmdOutputJobID := extractRayJobID(string(output))
+
+		rayJob := getAndCheckRayJob(namespace, rayJobName, cmdOutputJobID, "SUCCEEDED", "Complete")
+		Expect(rayJob.Spec.TTLSecondsAfterFinished).To(Equal(int32(0)))
+		Expect(rayJob.Spec.ShutdownAfterJobFinishes).To(BeTrue())
+	})
```
Add another testcase for `--ttl-seconds-after-finished` is not used at all.
